src/GoogleFit/FitOnFhir.GoogleFit.Tests/UsersServiceTests.cs

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using FitOnFhir.Common.Repositories;
+using FitOnFhir.Common.Tests;
+using FitOnFhir.Common.Tests.Mocks;
+using FitOnFhir.GoogleFit.Client.Models;
+using FitOnFhir.GoogleFit.Client.Responses;
+using FitOnFhir.GoogleFit.Common;
+using FitOnFhir.GoogleFit.Repositories;
+using FitOnFhir.GoogleFit.Services;
+using Microsoft.Extensions.Logging;
+using NSubstitute;
+using Xunit;
+
+namespace FitOnFhir.GoogleFit.Tests
+{
+    public class UsersServiceTests : UsersServiceBaseTests
+    {
+        private readonly IGoogleFitUserTableRepository _googleFitUserRepository;
+        private readonly IUsersKeyVaultRepository _usersKeyVaultRepository;
+        private readonly IGoogleFitAuthService _authService;
+        private readonly MockLogger<UsersService> _logger;
+        private readonly UsersService _usersService;
+
+        public UsersServiceTests()
+        {
+            _googleFitUserRepository = Substitute.For<IGoogleFitUserTableRepository>();
+            _usersKeyVaultRepository = Substitute.For<IUsersKeyVaultRepository>();
+            _authService = Substitute.For<IGoogleFitAuthService>();
+            _logger = Substitute.For<MockLogger<UsersService>>();
+
+            _usersService = new UsersService(
+                ResourceService,
+                UsersTableRepository,
+                _googleFitUserRepository,
+                _usersKeyVaultRepository,
+                _authService,
+                _logger);
+
+            // Default responses.
+            _authService.AuthTokensRequest(Arg.Any<string>(), Arg.Any<CancellationToken>()).Returns(Task.FromResult(Data.GetAuthTokensResponse()));
+        }
+
+        protected override Hl7.Fhir.Model.Bundle PatientBundle => Data.GetBundle(Data.GetPatient());
+
+        protected override Func<Task> ExecuteAuthorizationCallback => () => _usersService.ProcessAuthorizationCallback("TestAuthCode", CancellationToken.None);
+
+        protected override string ExpectedPatientIdentifierSystem => Data.Issuer;
+
+        protected override string ExpectedPatientId => Data.PatientId;
+
+        protected override string ExpectedPlatformUserId => Data.GoogleUserId;
+
+        protected override string ExpectedPlatform => GoogleFitConstants.GoogleFitPlatformName;
+
+        [Fact]
+        public async Task GivenAuthCodeIsNull_WhenProcessAuthorizationCallbackCalled_ConditionIsLogged()
+        {
+            await _usersService.ProcessAuthorizationCallback(null, CancellationToken.None);
+
+            _logger.Received(1).Log(
+                Arg.Is<LogLevel>(lvl => lvl == LogLevel.Information),
+                Arg.Is<string>(msg => msg == "ProcessAuthorizationCallback called with no auth code"));
```

Comment:
I've always been a little iffy on these tests that check for logger calls.  Logging can sometimes be very informal, and if we are testing for the exactness of the message logged itself, it seems quite possible that down the road someone may decide they want to change a logger line for debugging purposes in the moment.  In that case, tests like these will fail, when we shouldn't necessarily be holding a standard to the logging message itself.

Reply:
That is true. However, I did want some definitive way to test that this condition was actually handled appropriately. If the message is changed, it's probably a good idea that the test also needs to be updated.